components/raftstore/src/store/fsm/peer.rs

```diff
@@ -4683,17 +4701,72 @@ where
                 "epoch" => ?region_epoch,
                 "current_epoch" => ?region.get_region_epoch(),
             );
+            // test purpose
+            test_only_callback(cb, self.fsm.peer.region_buckets.clone());
             return;
         }
-        let mut region_buckets = Buckets::default();
-        region_buckets.version = timespec_to_ns(monotonic_raw_now());
-        region_buckets.set_keys(bucket_keys.into());
-        // add region's start/end key
-        region_buckets
-            .keys
-            .insert(0, region.get_start_key().to_vec());
-        region_buckets.keys.push(region.get_end_key().to_vec());
-        self.fsm.peer.region_buckets = region_buckets;
+
+        if let Some(bucket_ranges) = bucket_ranges {
+            assert_eq!(buckets.len(), bucket_ranges.len());
+            let mut region_buckets_keys = self.fsm.peer.region_buckets.keys.to_vec();
+            let mut region_buckets = Buckets::default();
+            region_buckets.version = timespec_to_ns(monotonic_raw_now());
+            let mut i = 0;
+            let mut j = 0;
+            let buckets = buckets.drain(..);
+            for mut bucket in buckets {
+                while i < region_buckets_keys.len() {
+                    if region_buckets_keys[i] == bucket_ranges[j].0 {
+                        // the bucket size is small and does not have split keys,
+                        // then it should be merged with its left neighbor
+                        if bucket.keys.is_empty()
+                            && bucket.size
+                                <= self.ctx.coprocessor_host.cfg.region_bucket_merge_size.0
+                        {
+                            // i is not the last entry (which is end key)
+                            if region_buckets_keys.len() > 2 {
+                                region_buckets_keys.remove(i); // bucket is too small
+                                // the key is removed, to compensate the += 1 below.
+                                i -= 1;
+                            }
+                        } else {
+                            // insert new bucket keys (split the original bucket)
+                            for bucket_key in bucket.keys.drain(..) {
+                                i += 1;
+                                region_buckets_keys.insert(i, bucket_key);
+                            }
+                        }
+                    }
+                    i += 1;
+                }
+                j += 1;
+            }
+
+            region_buckets.set_keys(region_buckets_keys.into());
+            assert_eq!(j, bucket_ranges.len());
+            self.fsm.peer.region_buckets = region_buckets;
+        } else {
+            assert_eq!(buckets.len(), 1);
+            let bucket_keys = buckets.pop().unwrap().keys;
+            let mut region_buckets = Buckets::default();
+            let now = monotonic_raw_now();
```

Comment:
Should use system time instead of monotontic_raw_now. The latter is an arbitrary number the kernel choose.

Reply:
Here all we need is a monotonic increasing number to differentiate Buckets right? System time can go backwards which can lead to confuse about which Bucket version is newer.

Reply:
It needs to increase monotonic across machines.

Reply:
cross machine comparison rely on region's epoch as today bucket version will compare epoch first.